### DISCUSSION TITLE
Can we use getStaticProps and getServerSideProps in a same page?

### DISCUSSION THREAD

**Discussion by IRediTOTO** [vercel/next.js#17649] [Ideas]
Use both of GSP and GSSP in page will not need call API anymore, this look so good I think :)

**Comment by mxmtsk** [comment]
Why do you need both in one page? The whole purpose of getStaticProps is to avoid server side rending completely on each request.

**Reply by IRediTOTO** [reply]
Example: I want to check the current user is bought an item to show "CANT BUY" on site, I must call api querry database to check. If can use getServerSideProps, we can check it on server :) of couse it should be load after all getStaticProps done and HTML sent to client. No more headache with API to check this check that on client.

**Comment by lfades** [comment]
TL;DR `getStaticProps` and `getServerSideProps` aren't compatible with each other.

With `getStaticProps` the output is completely static HTML, if you then add `getServerSideProps` then the page has to be generated at runtime, meaning whatever `getStaticProps` did no longer matters because you still need to generate the HTML at runtime.

**Reply by IRediTOTO** [reply]
I know about that. Just an idea. Maybe great people build this can consider. getStaticProps  render static  HTML send all to client first, after that getServerSideProps  do its job, send data to client and append data to static HTML like a additional data.

**Comment by kevva** [comment]
I think this is a good idea as well as an alternative/replacement for the `fallback: blocking` behaviour. Currently I have to sacrifice `GSP` in favour of `GSSP` because I use AMP and in some cases I need to access the query string, e.g. for filtering of products so I don't have to deal with the horrible `amp-list`, or access the `response` to be able to set the response code.

> TL;DR getStaticProps and getServerSideProps aren't compatible with each other.

I don't see why they would be mutually exclusive, since this basically is what `fallback: blocking` is doing in the background, except you'd give more control to the user.

**Reply by lfades** [reply]
@kevva What fallback blocking does is to run a serverless function if the page wasn't generated at build time, every request after that will get static HTML instead of running a serverless function again. The same thing applies here, a function runs because there's no static content, but it doesn't run if there's static content.

**Reply by kevva** [reply]
@lfades , exactly, I'm not strictly speaking of having `GSSP` and `GSP` side by side, but an API that exposes the `request` and `response`, or just specific parts like the query string to avoid possible side effects and keeping the API uniform, for the `fallback: blocking` behaviour in `GSP` when generating the static page. So yeah, it's not exactly the same as suggested by @IRediTOTO.

This is somewhat how I use `GSSP` right now, looking at the query string for filtering, sorting and pagination parameters, fetching data and then setting `s-maxage=1` with `stale-while-revalidate`. However, I have a lot of other URLs that share the same page in Next.js (I have a catch all route that looks up the page type from an API) that doesn't need this and would benefit from using `GSP` only, but due to `GSP` and `GSSP` being mutually exclusive, I have to opt in for one.
